### PR TITLE
feature: add go to definition for scaladoc

### PR DESCRIPTION
**Pull request by kasiaMarek**
https://github.com/scalameta/metals/issues/5736

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        else {
+          symbol.splitAt(indexOfDot + 1) match {
+            case ("this/", rest) => contextSymbols.withThis(rest)
+            case ("package/", rest) => contextSymbols.withPackage(rest)
+            case _ if symbol.contains("/") => List(symbol)
+            case _ => all
+          }
+        }
+
+      List(symbol.indexOf("("), symbol.indexOf("[")).filter(_ >= 0) match {
```
Comment (jkciesluk) [inline metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala:172]: Maybe something with
```scala
if (symbol.contains('(') || symbol.contains('[')) {
  val toTake = Math.max(symbol.indexOf('('), symbol.indexOf('['))
  withPrefixes.flatMap(sym => List(MethodSymbol(sym.take(toTake))))
else ...
```

**Review by jkciesluk — CHANGES_REQUESTED** [review]
Great work!
Don't we need to consider more cases in `loop`?

File: metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+      path: AbsolutePath,
+      params: TextDocumentPositionParams,
+  ): Option[DefinitionResult] = {
+    for {
+      buffer <- buffers.get(path)
+      position <- params.getPosition().toMeta(Input.String(buffer))
+      symbol <- extractScalaDocLinkAtPos(buffer, position)
+      contextSymbols = getContext(path, position)
+      scalaMetaSymbols = symbol.toScalaMetaSymbols(contextSymbols)
+      _ = scribe.info(
```
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala:37]: Should be debug.

File: metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        s"looking for definition for scaladoc symbol: $symbol considering alternatives: ${scalaMetaSymbols
+            .map(_.showSymbol)
+            .mkString(", ")}"
+      )
+      definitionResult <- scalaMetaSymbols.collectFirst { sym =>
```
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala:42]: I don't think so. According to scaladoc documentation if you have class and object it will prioritise the class. We do the same. Also we search: 1. this 2. this package 3. global, which is also compatible with how scopes work.

File: metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+      symbol <- ScalaDocLink.atOffset(comment.text, offset)
+    } yield symbol
+
+  private def getContext(
+      path: AbsolutePath,
+      pos: Position,
+  ): ContextSymbols = {
+    def extractName(ref: Term): String =
+      ref match {
+        case Term.Select(qual, name) => s"${extractName(qual)}/${name.value}"
```
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala:109]: For packages we need that too.

File: metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+      }
+    }
+
+  private def symbolWithFixedPackages = {
+    val fixedPackages =
+      ScalaDocLink
+        .splitAt(value, '.')
+        .map { str =>
+          if (
+            str.headOption.exists(_.isLower) ||
```
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala:228]: Yes, and those will be a problem. I'm not sure how to go about this. If we don't assume that packages start with lower case, we'd have to produce quite a lot of possible symbols. Maybe not that much, but I could easily get to ~50 to produce for possible scopes and types (class, object, method). I guess we could try a different approach and instead of producing the symbols create a fuzzy symbol search in `GlobalSymbolIndex`.

File: metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+      val optIndexOfDot =
+        ScalaDocLink.findIndicesOf(value, List('.')).headOption
+      def all = List(symbol) ++
+        contextSymbols.withThis(symbol) ++
+        contextSymbols.withPackage(symbol)
+      val withPrefixes: List[String] =
+        optIndexOfDot
+          .map { indexOfDot =>
+            symbol.splitAt(indexOfDot + 1) match {
+              case ("this/", rest) => contextSymbols.withThis(rest)
```
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala:182]: No, It'd be `this.<symbol>`. I'm searching on a symbol before substitution of `.` to `/` for packages and splitting on the other one. I see how that is confusing. I can try to rewrite this.

File: tests/unit/src/test/scala/tests/ScaladocSymbolsSuite.scala [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    List(
+      "`this.b`/`B.B`#",
+      "`this.b`/`B.B`.",
+      "`this.b`/`B.B`(+n).",
+    ),
+  )
+
+  check(
+    "escape2-scaladoc-link",
+    "this\\.B",
```
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply tests/unit/src/test/scala/tests/ScaladocSymbolsSuite.scala:51]: Now that you've mentioned it... It's a bit of an odd edge case but `\` can be used as escape, so to fully support this now I'm dropping `\` and wrapping it in backticks.

File: tests/unit/src/test/scala/tests/DefinitionLspSuite.scala [hunk 8]
```diff
@@ -606,4 +606,114 @@ class DefinitionLspSuite extends BaseLspSuite("definition") {
...
+      )
+      _ = client.messageRequests.clear()
+      _ <- server.didOpen("a/src/main/scala/a/Main.scala")
+      locations <- server.definition(
+        "a/src/main/scala/a/Main.scala",
+        testCase,
+        workspace,
+      )
+      _ = assert(locations.nonEmpty)
+      _ = assert(locations.head.getUri().endsWith("a/Main.scala"))
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply tests/unit/src/test/scala/tests/DefinitionLspSuite.scala:674]: What do you think about this?

File: metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/ScaladocDefinitionProvider.scala]: Let's get back to this as a follow back

**Review by tgodzik — APPROVED** [review]
Great work!

**Review by jkciesluk — APPROVED** [review]
LGTM
